### ISSUE TITLE
JSON Test Serialization on DELETE endpoints Regression in 5.1.4

### ISSUE THREAD

**Issue by brandonhilkert** [rails/rails#30570]
### Steps to reproduce

We tried to update to `5.1.4` and 2 tests broke because of the following...

Test app: https://github.com/brandonhilkert/bulk-destroy-test

Delete endpoints with supplied a supplied JSON payload are double serialized when using `as: :json` in the test.

I added a bulk endpoint here: https://github.com/brandonhilkert/bulk-destroy-test/commit/4cfb0030448aceb51f9197ff8e2f0ee2b7261a46

### Expected behavior

Test should pass

### Actual behavior

Test:
```ruby
  test "deletes all posts with ID" do
    assert_difference('Post.count', -1) do
      delete bulk_posts_url,
        params: { ids: [posts(:one).id, posts(:two).id] },
        as:     :json
    end
  end
```

```
(byebug) params
<ActionController::Parameters {"{\"ids\":"=>{"980190962,298486374"=>{"}"=>nil}}, "format"=>:json, "controller"=>"posts", "action"=>"bulk", "post"=>{}} permitted: false>
```

Removing `as: :json` causes the params to be properly serialized and test passes:

```
  test "deletes all posts with ID" do
    assert_difference('Post.count', -1) do
      delete bulk_posts_url,
        params: { ids: [posts(:one).id, posts(:two).id] }
    end
  end
```

```
(byebug) params
<ActionController::Parameters {"ids"=>["980190962", "298486374"], "format"=>:json, "controller"=>"posts", "action"=>"bulk"} permitted: false>
```

### System configuration
**Rails version**: 5.1.4

**Ruby version**: 2.4.1p111

**Comment by PHedkvist** [comment]
I believe you meant to put -2 in assert_difference. If I remove `as: :json` I get an error using destroy_all, I use `posts.each(&:destroy)` which passes the test. Regardless I get the same behaviour using `as: :json`.

**Comment by brandonhilkert** [comment]
`destroy` vs. `destroy_all` should make no different. You're right, it should be `-2`. I've changed it per your suggestions and the test still fails with `as: :json` and passes when `as: :json` is removed.

ref: https://github.com/brandonhilkert/bulk-destroy-test/commit/edf9aaad6d15928f3baead24470a09a109538fdc

**Comment by morgoth** [comment]
We have the same problem with 5.1.4.
We're not using `as: :json`, but set custom request headers like:
```
{
  "CONTENT_TYPE" => "application/json",
  "HTTP_ACCEPT" => "application/json"
}
```

**Comment by morgoth** [comment]
@brandonhilkert I found that this is caused by a change in `rack-test` https://github.com/rack-test/rack-test/issues/200 which was released in 0.7.0. You can verify it by using Rails 5.1.4 and rack-test 0.6.3

Rails allows 0.7.0 since https://github.com/rails/rails/pull/29859

**Comment by morgoth** [comment]
The issue is still present in the `rack-test` gem, that's still causing a regression in Rails

**Comment by y-yagi** [comment]
Perhaps this issue will be solved with https://github.com/rack-test/rack-test/pull/223.

**Comment by perlun** [comment]
@y-yagi @morgoth I just released rack-test 1.0.0 (yeah, :tada:) - please let me know if the problem persists.

**Comment by y-yagi** [comment]
Awesome :tada: 
I have confirmed that original issue fixed with rack-test 1.0.0. 
So I close this. But if anyone still has problem, please reopen it freely. Thanks!
